Add reusable widget to display single metric w/ sparkline & evolution percent (+ other changes)

diosmosis opened the pull request:
*This PR is not done but is ready for an initial review.*

Changes:
- Added new single-metric-value widget (angular based).
- Add query param filter_last_period_evolution that adds `_past`/`_evolution` metrics to any report. (not documented)
- Found an issue in `compileAngularComponents()`. The method uses the scope for an element, but in some cases `compileAngularComponents()` is called again on the same element (like when reloading a widget). This means that the scope is re-used when compiling a widget again, which in turn means its not possible to destroy angular components/directives. In the dashboard I forced the creation of a new scope, not sure if it's the best way to fix.
- Added `Report:: getMetricDocumentationForReport()` to access metric documentation for a report. Don't need it if an API.getProcessedReport, but that doesn't show the columns from `filter_last_period_evolution`.
- Added a `compileAngularComponents()` call to widget preview.
- Noticed when adding a widget it's reloaded several times (and all widgets w/ the same 'unique ID' are reloaded). Removed the extra reloads.

Note: if you'd rather the widget be rendered server side, let me know, just trying to use angular for all new work.

Fixes #12358 

CC @mattab / @tsteur for additional changes made + angular related fixes

![image](https://user-images.githubusercontent.com/125140/41830861-52905720-77f8-11e8-8417-fbc48740c083.png)

![image](https://user-images.githubusercontent.com/125140/41830879-7346f1a4-77f8-11e8-9c08-16bc93d52ba9.png)

diosmosis committed:
Add empty metric for single metric view.

diosmosis committed:
Add new isReusable property to widget metadata & if set to true, do not grey out the widget in the dashboard manager, even if the widget is used in the dashboard.

diosmosis committed:
Initial working version of single metric view.

diosmosis committed:
Get single metric view widget to work and look correctly (no series picker).

diosmosis committed:
Add series picker to single metric widget and add filter_last_period_evolution parameter.

diosmosis committed:
Persist metric change through dashboard widget parameter saving.

diosmosis committed:
Loading state for single metric view.

diosmosis committed:
Make new evolution param work on processed reports + tweak component implementation.

diosmosis committed:
Tweak CSS and make sure angular components are compiled in widget preview.

diosmosis committed:
Make component work with widget preview and avoid unnecessary widget reloads when multiple widgets of the same type are shown.

mattab commented on plugins/CoreHome/templates/_angularComponent.twig:
```diff
@@ -0,0 +1,5 @@
+<{{ componentName }}
+    {% for key, value in componentParameters %}
+    {{ key }}="{{ value|e('html') }}"
```
the componentName and key should be escaped here to be safe?

diosmosis replied:
👍 might be automatic w/ twig? can't remember, but will double check

mattab commented:
Feedback:

* It's usable and looks very nice to have only one widget with the metric picker :+1: 
* If we move the Sparklines rendering to client side, maybe we should aim to also convert the sparklines twig code to angular, as to avoid having the sparklines logic duplicated. Maybe this would be best done directly in this PR rather than later?
* some of the metadata is not displayed yet (percentage sign for rate metrics, currency for revenue metrics), 
* sometimes the metric is shown empty for some (instead of zero)
* is the `isReusable` flag needed? (the current/desired behavior is to allow adding the same widget multiple times for all widgets)
* left a code comment re: maybe missing escaping

diosmosis commented:
> is the isReusable flag needed?

It's used to display the 'Metric' widget w/ black text instead of grey text. Not required, I guess, but I think it's slightly better UX.

diosmosis committed:
Generalize JS lastN range period computing and use to create standalone sparkline angular component and get rid of need for "past-period" argument to single metric view.

diosmosis committed:
Add format_metrics: "1" to API.get method.

diosmosis commented:
> sometimes the metric is shown empty for some (instead of zero)

Do you remember which metric? Not currently able to reproduce.

diosmosis committed:
Add escaping to _angularComponent.twig.

diosmosis commented:
> some of the metadata is not displayed yet (percentage sign for rate metrics, currency for revenue metrics),

Some of this is hard to fix. Processed metrics (like percents) I could display via `format_metrics=1`, but normal metrics like `revenue` are not currently formatted in API output. I will add a fix in another PR.

diosmosis committed:
Merge branch '3.x-dev' into 12358-single-metric-widget

diosmosis committed:
hacky fix for formatting revenue columns

diosmosis commented:
@mattab fixed the revenue issue w/ a hacky fix. Needs review, will be adding tests now.

diosmosis committed:
Format past data values & allow evolution to be calculated for processed metrics.

diosmosis committed:
Merge branch '3.x-dev' into 12358-single-metric-widget

diosmosis committed:
filter evolution changes

diosmosis committed:
Fix issue in subtable recursion for processed metric computation & metric formatting + add new processed metric compute hooks to fix bug in evolution calculation on subtables.

diosmosis commented:
Found a couple issues in processed metrics calculation & metric formatting when getting the evolution filter to work:

- The recursion on subtables would recurse multiple times, once per metric. Now it does this once per table.
- Processed metrics were not computed in any order, so processed metrics couldn't use other processed metrics. Fixed by using getDependentMetrics() to determine order of metrics.
- EvolutionMetric did not work correctly on subtables since it would always use the root past datatable when computing evolutions for subtables. Fixed by adding two new non-API hooks to ProcessedMetric: `beforeComputeSubtable()`, `afterComputeSubtable()`

Also noticed that new processed metrics will never appear in processed report output because the columns are limited to the metrics/processed metrics in the Report class. This makes me think it might be useful to make the Report class mutable, so DataTablePostProcessor (for instance) can add new processed metrics. What do you think @tsteur / @mattab?

tsteur commented on core/Widget/WidgetConfig.php:
```diff
@@ -366,4 +367,31 @@ public function isWide()
+     *
+     * @return bool
+     */
+    public function isReusable()
```
Is that new setting really needed? IMO adding any widget multiple times can always make sense as you may want to add it to dashboard multiple times with different visualizations or sort orders etc. 

If we keep it: `reusable` is maybe not best wording maybe as it is pretty much only useful for dashboard in this is not quite clear. 

I'd even rather hard code this in the dashboard JS instead of adding this new setting.

tsteur replied:
Personally, it be fine to always remove the "grey" color and show it in regular color (I presume @mattab  will want to keep this functionality). If that's not fine and we still need to show it grey, I would rather hard code it for this widget in the dashboard then adding this new API. Eg based on the "Generic" category or widget unique Id or something.

diosmosis replied:
👍

mattab replied:
It's not a 100% requirement and sounds good to hardcode it. It's quite useful to show the already-added-widgets differently, as people may forget the widget is already added in the bottom of the dashboard.

tsteur commented on core/API/DataTablePostProcessor.php:
```diff
@@ -421,16 +430,68 @@ public function computeProcessedMetrics(DataTable $dataTable)
+            }
+        }
+
+        foreach ($dataTable->getRows() as $row) {
```
can you comment here why this is needed and what it does? 

Also can we check this for some bigger subtables re performance etc?

diosmosis replied:
Note: this is one of several issues I fixed while doing this work. Before this loop was executed for every processed metric being computed, instead of once for the whole table. So it was doing extra work.

diosmosis committed:
remove isReusable property.

diosmosis committed:
attempting to change strategy

diosmosis committed:
simpler solution that does not require backend changes

diosmosis committed:
remove unneeded code + fix issue w/ formatted metrics

diosmosis committed:
remove some more unneeded code

diosmosis commented:
@tsteur FYI I changed the strategy here to calculate the evolution data on in the component so the new parameter isn't required. There's less backend changes now, the changes that exist are mostly fixes for issues I found.

diosmosis committed:
write UI test

diosmosis committed:
add new screenshots

diosmosis committed:
Merge branch '3.x-dev' into 12358-single-metric-widget

tsteur commented:
should be fine @diosmosis 👍

mattab commented:
@diosmosis Looks good overall, a awesome new feature :+1: 

Feedback:

* KPI metrics for each Goal (Goal conversions, Conversion rate, Revenue). Currently it only has the "overall conversions/revenue. 
* the `overall conversion rate (visits with a completed goal)` is missing, not sure if it's expected but it would be good to have 
* in general when comparing the list of metrics available in the metric picker VS all the metrics listed in https://github.com/innocraft/plugin-CustomReports/pull/56 -> how would it be possible to have more/all metrics available in this widget picker like they are in Custom Reports?
* move the "Generic" category above the "Diagnostic" category if possible

diosmosis commented:
> KPI metrics for each Goal (Goal conversions, Conversion rate, Revenue). Currently it only has the "overall conversions/revenue.

The series picker doesn't let you pick more than a metric (ie, you can't pick a metric + an idGoal). I'll put another select for goals when a goal metric is displayed.

> in general when comparing the list of metrics available in the metric picker VS all the metrics listed in innocraft/plugin-CustomReports#56 -> how would it be possible to have more/all metrics available in this widget picker like they are in Custom Reports?

This widget just uses the `API.get` report. Some metrics used in custom reports are probably metrics that aren't saved "standalone" (ie, maybe they're always w/ a specific report?). Ideally we would add these metrics to `API.get` to show them.

diosmosis commented:
@mattab I can probably hack something in like that, but it would create a very large series picker.

mattab commented:
It will be OK to have a large picker as long as it's not buggy. in the future, we'll improve the metric picker so it looks like the custom reports metric picker.

diosmosis committed:
Add all goals to single metric view picker.

diosmosis commented:
@mattab added goal metrics, deserves more tests since it's a hacky change. (did not modify the API)

diosmosis committed:
move category

diosmosis commented:
Moved the category:

![image](https://user-images.githubusercontent.com/125140/43500117-c8146b9e-9503-11e8-93b5-4ee4cada8120.png)

diosmosis committed:
Merge branch '3.x-dev' into 12358-single-metric-widget

diosmosis committed:
fix test

diosmosis commented:
@mattab approved on slack, merging after fixing build

diosmosis committed:
fixing more tests

diosmosis committed:
Fixing some UI tests.

diosmosis committed:
Update more screenshots.

diosmosis committed:
update two more screenshots
